gandalf.js

```diff
@@ -1,67 +1,64 @@
-/**
```

Comment:
SCRIPT-This function seems to have grown quite complex. Can we break it down further?